include/pybind11/cast.h

```diff
@@ -1055,6 +1059,16 @@ T cast(const handle &handle) {
     return T(reinterpret_borrow<object>(handle));
 }
 
+// Note that `cast<PyObject *>(obj)` increments the reference count of `obj`.
+// This is necessary for the case that `obj` is a temporary.
```

Comment:
Is this actually necessary? Seems a bit unintuitive here. When are situations when obj would be a temporary?

Reply:
In a callback (functional.h):

* https://github.com/pybind/pybind11/blob/7ab88d2e4fbb7fb21473f87b17b68a1f0e51cb4e/include/pybind11/functional.h#L109

That is using `object::cast<PyObject *>()` which winds up in `py::cast<PyObject *>(handle)`:

* https://github.com/pybind/pybind11/blob/7ab88d2e4fbb7fb21473f87b17b68a1f0e51cb4e/include/pybind11/cast.h#L1136-L1139

This actually took me a while to discover and debug.

I don't see an easy way to make the code in functional.h behave differently.

But I'm also thinking:
```
cast<PyObject *>(something_that_generates_a_temporary_object())
```
with the temporary passed as `handle` is a really bad trap in general (especially if the `cast` is actually written as `cast<T>` buried in the guts of some library).

By adopting the `inc_ref()` convention we don't have to do anything special for callbacks and we avoid the trap.

Users have to learn once that the refcount is incremented, it always applies.

This is only for users that find it necessary to actually drop down to the level of dealing with raw `PyObject *`s. In that context, having to be super mindful about refcounts is a given, and careful leak checking of the tests is a standard practice for me at least (`while True:` & top is the most reliable btw).

Another way to look at it:

* Risk of a accidental leaks (with the `inc_ref()`).
vs
* Risk of UB (segfault if you're lucky).

What tipped the scale in my mind was that the `inc_ref()` convention means nothing else needs to change, which is usually a good sign for changes like this (fitting something slightly different into an existing system).

Reply:
> This is necessary incase obj is a temporary

Which is why I am advocating for the additional overload below. For C++ calls, we can determine if the object is a temporary (ie. it's an rvalue) and avoid having to increment it under those circumstances.